etcdserver/server_test.go

```diff
@@ -823,6 +826,94 @@ func TestTriggerSnap(t *testing.T) {
 	}
 }
 
+// TestConcurrentApplyAndSnapshot will send out snapshots concurrently with
+// proposals.
+func TestConcurrentApplyAndSnapshot(t *testing.T) {
+	n := newReadyNode()
+	cl := newCluster("abc")
+	cl.SetStore(store.New())
+
+	testdir, err := ioutil.TempDir(os.TempDir(), "testsnapdir")
+	if err != nil {
+		t.Fatalf("Couldn't open tempdir (%v)", err)
+	}
+	defer os.RemoveAll(testdir)
+	if err := os.MkdirAll(testdir+"/member/snap", 0755); err != nil {
+		t.Fatalf("Couldn't make snap dir (%v)", err)
+	}
+
+	rs := raft.NewMemoryStorage()
+	tr := newSnapTransporter(testdir)
+	s := &EtcdServer{
+		cfg: &ServerConfig{
+			V3demo:  true,
+			DataDir: testdir,
+		},
+		r: raftNode{
+			Node:        n,
+			transport:   tr,
+			storage:     &storageRecorder{dbPath: testdir},
+			raftStorage: rs,
+		},
+		store:    cl.store,
+		cluster:  cl,
+		msgSnapC: make(chan raftpb.Message, 100/3),
+	}
+
+	s.kv = dstorage.New(
+		path.Join(testdir, "testdb.db"),
+		&s.consistIndex)
+
+	s.start()
+
+	// submit applied entries and snap entries
+	idx := uint64(0)
+	outdated := 0
+	accepted := 0
+	for k := 1; k < 50; k++ {
+		idx++
+		ch := s.w.Register(uint64(idx))
+		req := &pb.Request{Method: "QGET", ID: uint64(idx)}
+		ent := raftpb.Entry{Index: uint64(idx), Data: pbutil.MustMarshal(req)}
+		ready := raft.Ready{Entries: []raftpb.Entry{ent}}
+		n.readyc <- ready
+
+		ready = raft.Ready{CommittedEntries: []raftpb.Entry{ent}}
+		n.readyc <- ready
+
+		// "idx" applied
+		<-ch
```

Comment:
this makes apply and snapshot actually not concurrent. 

The test failed without this patch probably because ep failed to be updated after apply before doing the snapshot.

```
case ep = <-etcdprogc:
case m := <-s.msgSnapC:
```